File: api/app/protection.py
"""Protection des données : couverture, fraîcheur et risques de sauvegarde.

On agrège trois sources — Proxmox Backup Server, les vzdump d'un PVE et les
tâches Hyper Backup d'un DSM — pour répondre à deux questions : qu'est-ce qui
n'est pas sauvegardé, et quelles sauvegardes ne sont plus fiables.
"""
from __future__ import annotations

import asyncio
import datetime as dt
import logging
from typing import Any

from .bus import bus
from .collectors import pbs as pbs_col
from .collectors import proxmox as pve_col
from .collectors import synology as syno_col
from .db import fetch_all

log = logging.getLogger("mba.protection")

# Au-delà, une sauvegarde n'est plus considérée comme fraîche.
STALE_HOURS = 36
CRITICAL_HOURS = 24 * 7
# Un datastore trop plein fait échouer les sauvegardes suivantes.
STORE_WARN = 80.0
STORE_CRIT = 92.0


def _age_hours(timestamp: Any) -> float | None:
    """Âge en heures d'un epoch (secondes) ou d'une date ISO."""
    if timestamp in (None, "", 0):
        return None
    try:
        if isinstance(timestamp, (int, float)):
            when = dt.datetime.fromtimestamp(float(timestamp), dt.timezone.utc)
        else:
            when = dt.datetime.fromisoformat(str(timestamp).replace("Z", "+00:00"))
            if when.tzinfo is None:
                when = when.replace(tzinfo=dt.timezone.utc)
    except (ValueError, OSError, OverflowError):
        return None
    return (dt.datetime.now(dt.timezone.utc) - when).total_seconds() / 3600.0


def _freshness(age: float | None) -> str:
    if age is None:
        return "unknown"
    if age > CRITICAL_HOURS:
        return "critical"
    if age > STALE_HOURS:
        return "stale"
    return "fresh"


# ------------------------------------------------------------------ collecte
async def _from_pbs(host: dict) -> dict[str, Any]:
    client = await pbs_col.client_for_host(host)
    try:
        snap = await client.snapshot()
    finally:
        await client.close()

    protected = []
    for group in snap["groups"]:
        age = _age_hours(group.get("last_backup"))
        protected.append({
            "source": "pbs",
            "source_host": host["name"],
            "source_host_id": host["id"],
            "name": f"{group['backup_type']}/{group['backup_id']}",
            "kind": group["backup_type"],
            "ref": str(group["backup_id"]),
            "store": group["store"],
            "count": group["count"],
            "last_backup": group.get("last_backup"),
            "age_hours": age,
            "freshness": _freshness(age),
            "owner": group.get("owner"),
        })
    return {
        "protected": protected,
        "datastores": [{**store, "source_host": host["name"], "source": "pbs"}
                       for store in snap["datastores"]],
        "tasks": [{**task, "source_host": host["name"], "source": "pbs"}
                  for task in snap["tasks"]],
        "notices": [],
    }


async def _from_pve(host: dict) -> dict[str, Any]:
    """Sauvegardes vzdump vues depuis l'hyperviseur."""
    client = await pve_col.client_for_host(host)
    protected: list[dict] = []
    try:
        nodes = (host.get("meta") or {}).get("nodes") or []
        if not nodes:
            fetched = await client.get("/nodes") or []
            nodes = [n["node"] for n in fetched]
        latest: dict[int, dict] = {}
        for node in nodes:
            for backup in await client.backups(node):
                vmid = backup.get("vmid")
                if vmid is None:
                    continue
                current = latest.get(vmid)
                if not current or (backup.get("created") or 0) > (current.get("created") or 0):
                    latest[vmid] = backup
        for vmid, backup in latest.items():
            age = _age_hours(backup.get("created"))
            protected.append({
                "source": "vzdump",
                "source_host": host["name"],
                "source_host_id": host["id"],
                "name": f"vzdump {vmid}",
                "kind": "guest",
                "ref": str(vmid),
                "store": backup.get("storage"),
                "count": 1,
                "size": backup.get("size"),
                "last_backup": backup.get("created"),
                "age_hours": age,
                "freshness": _freshness(age),
            })
    finally:
        await client.close()
    return {"protected": protected, "datastores": [], "tasks": [], "notices": []}


async def _from_synology(host: dict) -> dict[str, Any]:
    client = await syno_col.client_for_host(host)
    try:
        result = await client.hyper_backup()
    finally:
        await client.logout()

    tasks = result.get("tasks") or []
    notices: list[dict[str, Any]] = []
    if result.get("reason"):
        notices.append({
            "host": host["name"],
            "kind": "synology",
            "level": "info" if not result.get("installed") else "warning",
            "message": result["reason"],
        })

    protected, jobs = [], []
    for task in tasks:
        age = _age_hours(task.get("last_backup"))
        ok = str(task.get("last_result", "")).lower() in ("", "success", "done", "0", "normal")
        protected.append({
            "source": "hyperbackup",
            "source_host": host["name"],
            "source_host_id": host["id"],
            "name": task.get("name") or f"tâche {task.get('id')}",
            "kind": "hyperbackup",
            "ref": str(task.get("id")),
            "store": task.get("target"),
            "count": 1,
            "size": task.get("size"),
            "last_backup": task.get("last_backup"),
            "age_hours": age,
            "freshness": _freshness(age),
            "last_result": task.get("last_result"),
            "enabled": task.get("enabled", True),
            "is_c2": bool(task.get("is_c2")),
        })
        jobs.append({
            "source": "c2" if task.get("is_c2") else "hyperbackup",
            "source_host": host["name"],
            "type": "hyperbackup",
            "target": task.get("name"),
            "status": "OK" if ok else str(task.get("last_result")),
            "started": task.get("last_backup"),
            "ended": task.get("last_backup"),
            "schedule": task.get("schedule"),
            "next": task.get("next_backup"),
        })
    # Une tâche vers C2 est une sauvegarde hors site : on la distingue.
    for item in protected:
        if item.pop("is_c2", False):
            item["source"] = "c2"
            item["offsite"] = True

    return {"protected": protected, "datastores": [], "tasks": jobs, "notices": notices}


# ------------------------------------------------------------------ analyse
def _find_gaps(protected: list[dict], guests: list[dict], hosts: list[dict]) -> list[dict]:
    """Ce qui existe mais n'apparaît dans aucune sauvegarde."""
    covered_refs = {p["ref"] for p in protected}
    covered_names = {(p.get("name") or "").lower() for p in protected}
    gaps: list[dict] = []

    for guest in guests:
        vmid = str(guest.get("vmid"))
        if vmid in covered_refs:
            continue
        gaps.append({
            "kind": "guest",
            "name": guest.get("name"),
            "ref": vmid,
            "detail": f"{'VM' if guest.get('type') == 'qemu' else 'Conteneur'} {vmid} sur "
                      f"{guest.get('node')}",
            "host_id": guest.get("host_id"),
            "severity": "high" if guest.get("status") == "running" else "medium",
        })

    for host in hosts:
        # Un NAS se sauvegarde via Hyper Backup, un hyperviseur via ses invités :
        # on ne signale que les machines réellement laissées de côté.
        if host["kind"] not in ("linux", "docker"):
            continue
        # L'hôte Docker interne de MBA n'est pas une machine à sauvegarder.
        if (host.get("meta") or {}).get("builtin"):
            continue
        if host["name"].lower() in covered_names:
            continue
        if any(host["name"].lower() in (p.get("name") or "").lower() for p in protected):
            continue
        gaps.append({
            "kind": "host",
            "name": host["name"],
            "ref": str(host["id"]),
            "detail": f"{host['address']} — aucune sauvegarde identifiée",
            "host_id": host["id"],
            "severity": "medium",
        })
    return gaps


def _find_risks(protected: list[dict], datastores: list[dict], tasks: list[dict]) -> list[dict]:
    risks: list[dict] = []

    for item in protected:
        if item["freshness"] == "critical":
            risks.append({
                "severity": "critical",
                "title": f"{item['name']} : sauvegarde vieille de {item['age_hours'] / 24:.0f} jours",
                "detail": f"Dernière copie sur {item.get('store') or item['source_host']}.",
                "remediation": "Vérifie que la tâche tourne encore, puis relance-la.",
                "target": item["name"],
                "source": item["source"],
            })
        elif item["freshness"] == "stale":
            risks.append({
                "severity": "high",
                "title": f"{item['name']} : pas de sauvegarde depuis {item['age_hours']:.0f} h",
                "detail": f"Seuil de fraîcheur fixé à {STALE_HOURS} h.",
                "remediation": "Contrôle la planification de la tâche.",
                "target": item["name"],
                "source": item["source"],
            })
        elif item["freshness"] == "unknown":
            risks.append({
                "severity": "medium",
                "title": f"{item['name']} : date de dernière sauvegarde inconnue",
                "detail": "La source ne remonte pas d'horodatage exploitable.",
                "remediation": "Ouvre la tâche côté PBS ou DSM pour confirmer son état.",
                "target": item["name"],
                "source": item["source"],
            })

        if item.get("count", 0) == 1 and item["source"] == "pbs":
            risks.append({
                "severity": "medium",
                "title": f"{item['name']} : une seule version conservée",
                "detail": "Sans historique, une corruption propagée n'est plus rattrapable.",
                "remediation": "Augmente la rétention (prune) pour garder plusieurs points.",
                "target": item["name"],
                "source": item["source"],
            })

        if item.get("enabled") is False:
            risks.append({
                "severity": "high",
                "title": f"{item['name']} : tâche désactivée",
                "detail": "La planification existe mais ne s'exécute plus.",
                "remediation": "Réactive la tâche dans Hyper Backup.",
                "target": item["name"],
                "source": item["source"],
            })

    for store in datastores:
        if store["percent"] >= STORE_CRIT:
            risks.append({
                "severity": "critical",
                "title": f"Datastore {store['name']} rempli à {store['percent']:.0f} %",
                "detail": f"{store.get('source_host', '')} — les prochaines sauvegardes échoueront.",
                "remediation": "Lance un prune/GC, ou ajoute de la capacité.",
                "target": store["name"],
                "source": store.get("source", "pbs"),
            })
        elif store["percent"] >= STORE_WARN:
            risks.append({
                "severity": "medium",
                "title": f"Datastore {store['name']} à {store['percent']:.0f} %",
                "detail": store.get("estimated_full")
                and f"Saturation estimée : {store['estimated_full']}" or "",
                "remediation": "Anticipe un prune ou une extension.",
                "target": store["name"],
                "source": store.get("source", "pbs"),
            })

    unverified = [p for p in protected if p["source"] == "pbs" and not p.get("verified")]
    if unverified:
        risks.append({
            "severity": "medium",
            "title": f"{len(unverified)} groupe(s) sans vérification PBS",
            "detail": "Une sauvegarde jamais vérifiée peut être illisible au moment critique.",
            "remediation": "Planifie un job de vérification dans PBS.",
            "target": "vérification",
            "source": "pbs",
        })

    failed = [t for t in tasks
              if t.get("status") not in ("OK", "running", None) and t.get("status")]
    for task in failed[:10]:
        risks.append({
            "severity": "high",
            "title": f"Tâche en échec : {task.get('target') or task.get('type')}",
            "detail": f"{task.get('source_host', '')} — statut « {task.get('status')} ».",
            "remediation": "Consulte le journal de la tâche pour la cause exacte.",
            "target": task.get("target") or "",
            "source": task.get("source", "pbs"),
        })

    order = {"critical": 0, "high": 1, "medium": 2, "low": 3}
    risks.sort(key=lambda r: order.get(r["severity"], 9))
    return risks


# ---------------------------------------------------------------- agrégation
async def overview() -> dict[str, Any]:
    hosts = await fetch_all("SELECT * FROM hosts WHERE enabled")
    by_kind: dict[str, list[dict]] = {}
    for host in hosts:
        by_kind.setdefault(host["kind"], []).append(host)

    sources = [
        *[(_from_pbs, h) for h in by_kind.get("pbs", [])],
        *[(_from_pve, h) for h in by_kind.get("proxmox", [])],
        *[(_from_synology, h) for h in by_kind.get("synology", [])],
    ]
    results = await asyncio.gather(*(fn(host) for fn, host in sources), return_exceptions=True)

    protected: list[dict] = []
    datastores: list[dict] = []
    tasks: list[dict] = []
    errors: list[dict] = []
    notices: list[dict] = []
    for (fn, host), result in zip(sources, results):
        if isinstance(result, Exception):
            errors.append({"host": host["name"], "kind": host["kind"], "error": str(result)[:200]})
            log.info("Protection : %s indisponible (%s)", host["name"], result)
            continue
        protected.extend(result["protected"])
        datastores.extend(result["datastores"])
        tasks.extend(result["tasks"])
        notices.extend(result.get("notices") or [])

    # Les invités Proxmox constituent la population à couvrir.
    guests: list[dict] = []
    live = bus.latest("metrics.")
    for host in by_kind.get("proxmox", []):
        sample = live.get(f"metrics.{host['id']}", {})
        for guest in sample.get("guests") or []:
            guests.append({**guest, "host_id": host["id"]})

    gaps = _find_gaps(protected, guests, hosts)
    risks = _find_risks(protected, datastores, tasks)

    # Règle 3-2-1 : sans copie hors site, un sinistre local emporte tout.
    offsite = [p for p in protected if p.get("offsite")]
    if protected and not offsite:
        risks.insert(0, {
            "severity": "high",
            "title": "Aucune sauvegarde hors site",
            "detail": "Toutes les copies connues restent sur le même site que les données.",
            "remediation": "Ajoute une destination distante : Synology C2, un PBS déporté, "
                           "ou un stockage objet via Hyper Backup.",
            "target": "hors site",
            "source": "global",
        })

    fresh = len([p for p in protected if p["freshness"] == "fresh"])
    total_targets = len(protected) + len(gaps)
    return {
        "protected": sorted(protected, key=lambda p: (p["freshness"] != "fresh", p["name"])),
        "gaps": gaps,
        "risks": risks,
        "datastores": datastores,
        "tasks": sorted(tasks, key=lambda t: t.get("started") or 0, reverse=True)[:40],
        "errors": errors,
        "notices": notices,
        "summary": {
            "offsite": len(offsite),
            "protected": len(protected),
            "fresh": fresh,
            "stale": len([p for p in protected if p["freshness"] == "stale"]),
            "critical": len([p for p in protected if p["freshness"] == "critical"]),
            "gaps": len(gaps),
            "risks": len(risks),
            "risks_critical": len([r for r in risks if r["severity"] == "critical"]),
            # Part des objets connus qui disposent d'une sauvegarde récente.
            "coverage": round(100.0 * fresh / total_targets, 1) if total_targets else 100.0,
            "sources": len(sources),
            "total_size": sum(p.get("size") or 0 for p in protected),
        },
    }

File: ios/MBA/Core/Models/Protection.swift
import Foundation

/// Réponse de `GET /protection` : couverture des sauvegardes, écarts et risques,
/// agrégés sur toutes les sources connues (PBS, vzdump, Hyper Backup).
struct ProtectionOverview: Codable, Sendable {
    var protectedItems: [ProtectedItem]
    var gaps: [ProtectionGap]
    var risks: [ProtectionRisk]
    var datastores: [Datastore]
    var tasks: [BackupTask]
    var errors: [SourceError]
    var notices: [SourceNotice]
    var summary: Summary

    enum CodingKeys: String, CodingKey {
        case gaps, risks, datastores, tasks, errors, notices, summary
        case protectedItems = "protected"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        protectedItems = (try? container.decode([ProtectedItem].self, forKey: .protectedItems)) ?? []
        gaps = (try? container.decode([ProtectionGap].self, forKey: .gaps)) ?? []
        risks = (try? container.decode([ProtectionRisk].self, forKey: .risks)) ?? []
        datastores = (try? container.decode([Datastore].self, forKey: .datastores)) ?? []
        tasks = (try? container.decode([BackupTask].self, forKey: .tasks)) ?? []
        errors = (try? container.decode([SourceError].self, forKey: .errors)) ?? []
        notices = (try? container.decode([SourceNotice].self, forKey: .notices)) ?? []
        summary = (try? container.decode(Summary.self, forKey: .summary)) ?? Summary()
    }

    struct Summary: Codable, Sendable {
        var offsite: Int = 0
        var protectedCount: Int = 0
        var fresh: Int = 0
        var stale: Int = 0
        var critical: Int = 0
        var gaps: Int = 0
        var risks: Int = 0
        var risksCritical: Int = 0
        var coverage: Double = 100
        var sources: Int = 0
        var totalSize: Double = 0

        enum CodingKeys: String, CodingKey {
            case offsite, fresh, stale, critical, gaps, risks, coverage, sources
            case protectedCount = "protected"
            case risksCritical = "risks_critical"
            case totalSize = "total_size"
        }

        init() {}

        init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            offsite = (try? container.decode(Int.self, forKey: .offsite)) ?? 0
            protectedCount = (try? container.decode(Int.self, forKey: .protectedCount)) ?? 0
            fresh = (try? container.decode(Int.self, forKey: .fresh)) ?? 0
            stale = (try? container.decode(Int.self, forKey: .stale)) ?? 0
            critical = (try? container.decode(Int.self, forKey: .critical)) ?? 0
            gaps = (try? container.decode(Int.self, forKey: .gaps)) ?? 0
            risks = (try? container.decode(Int.self, forKey: .risks)) ?? 0
            risksCritical = (try? container.decode(Int.self, forKey: .risksCritical)) ?? 0
            coverage = (try? container.decode(Double.self, forKey: .coverage)) ?? 100
            sources = (try? container.decode(Int.self, forKey: .sources)) ?? 0
            totalSize = (try? container.decode(Double.self, forKey: .totalSize)) ?? 0
        }
    }
}

/// Un objet effectivement sauvegardé quelque part.
struct ProtectedItem: Codable, Identifiable, Hashable, Sendable {
    var source: BackupSource
    var sourceHost: String
    var sourceHostID: Int?
    var name: String
    var kind: String
    var ref: String
    var store: String?
    var count: Int?
    var size: Double?
    var lastBackup: Date?
    var ageHours: Double?
    var freshness: Freshness
    var enabled: Bool?
    var offsite: Bool
    var lastResult: String?

    var id: String { "\(sourceHost):\(source.rawValue):\(ref):\(name)" }

    enum CodingKeys: String, CodingKey {
        case source, name, kind, ref, store, count, size, freshness, enabled, offsite
        case sourceHost = "source_host"
        case sourceHostID = "source_host_id"
        case lastBackup = "last_backup"
        case ageHours = "age_hours"
        case lastResult = "last_result"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        source = (try? container.decode(BackupSource.self, forKey: .source)) ?? .other
        sourceHost = (try? container.decode(String.self, forKey: .sourceHost)) ?? "—"
        sourceHostID = try container.decodeIfPresent(Int.self, forKey: .sourceHostID)
        name = (try? container.decode(String.self, forKey: .name)) ?? "—"
        kind = (try? container.decode(String.self, forKey: .kind)) ?? ""
        ref = (try? container.decode(String.self, forKey: .ref)) ?? ""
        store = try container.decodeIfPresent(String.self, forKey: .store)
        count = try container.decodeIfPresent(Int.self, forKey: .count)
        size = try container.decodeIfPresent(Double.self, forKey: .size)
        lastBackup = container.decodeLooseDate(forKey: .lastBackup)
        ageHours = try container.decodeIfPresent(Double.self, forKey: .ageHours)
        freshness = (try? container.decode(Freshness.self, forKey: .freshness)) ?? .unknown
        enabled = try container.decodeIfPresent(Bool.self, forKey: .enabled)
        offsite = (try? container.decode(Bool.self, forKey: .offsite)) ?? false
        lastResult = try container.decodeIfPresent(String.self, forKey: .lastResult)
    }

    /// Là où la copie est posée : le datastore quand il est connu, sinon la
    /// machine qui la détient.
    var location: String { store?.isEmpty == false ? store! : sourceHost }

    func matches(_ query: String) -> Bool {
        guard !query.isEmpty else { return true }
        return [name, store, sourceHost]
            .compactMap { $0?.lowercased() }
            .contains { $0.contains(query) }
    }
}

/// Fraîcheur d'une sauvegarde, telle que le serveur la juge (36 h / 7 j).
enum Freshness: String, Codable, Sendable {
    case fresh, stale, critical, unknown

    var label: String {
        switch self {
        case .fresh: "à jour"
        case .stale: "en retard"
        case .critical: "périmée"
        case .unknown: "inconnue"
        }
    }

    var symbol: String {
        switch self {
        case .fresh: "checkmark.circle.fill"
        case .stale: "clock.badge.exclamationmark.fill"
        case .critical: "exclamationmark.octagon.fill"
        case .unknown: "questionmark.circle.fill"
        }
    }
}

/// Origine d'une sauvegarde ou d'un risque.
enum BackupSource: String, Codable, Sendable {
    case pbs, vzdump, hyperbackup, c2, global, other

    init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = Self(rawValue: raw) ?? .other
    }

    var label: String {
        switch self {
        case .pbs: "Proxmox Backup Server"
        case .vzdump: "vzdump (PVE)"
        case .hyperbackup: "Hyper Backup"
        case .c2: "Synology C2"
        case .global: "Analyse globale"
        case .other: "Autre"
        }
    }

    var shortLabel: String {
        switch self {
        case .pbs: "PBS"
        case .vzdump: "vzdump"
        case .hyperbackup: "Hyper Backup"
        case .c2: "C2"
        case .global: "Global"
        case .other: "Autre"
        }
    }
}

/// Un objet connu qui n'apparaît dans aucune sauvegarde.
struct ProtectionGap: Codable, Identifiable, Hashable, Sendable {
    var kind: String
    var name: String
    var ref: String
    var detail: String?
    var hostID: Int?
    var severity: FindingSeverity

    var id: String { "\(kind):\(ref)" }

    enum CodingKeys: String, CodingKey {
        case kind, name, ref, detail, severity
        case hostID = "host_id"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        kind = (try? container.decode(String.self, forKey: .kind)) ?? ""
        name = (try? container.decode(String.self, forKey: .name)) ?? "—"
        ref = (try? container.decode(String.self, forKey: .ref)) ?? ""
        detail = try container.decodeIfPresent(String.self, forKey: .detail)
        hostID = try container.decodeIfPresent(Int.self, forKey: .hostID)
        severity = (try? container.decode(FindingSeverity.self, forKey: .severity)) ?? .medium
    }

    var isGuest: Bool { kind == "guest" }
    var kindLabel: String { isGuest ? "VM / LXC" : "Machine" }
}

/// Un risque identifié sur la chaîne de sauvegarde.
struct ProtectionRisk: Codable, Identifiable, Hashable, Sendable {
    var severity: FindingSeverity
    var title: String
    var detail: String?
    var remediation: String?
    var target: String?
    var source: BackupSource

    var id: String { "\(severity.rawValue):\(title)" }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        severity = (try? container.decode(FindingSeverity.self, forKey: .severity)) ?? .medium
        title = (try? container.decode(String.self, forKey: .title)) ?? "—"
        detail = try container.decodeIfPresent(String.self, forKey: .detail)
        remediation = try container.decodeIfPresent(String.self, forKey: .remediation)
        target = try container.decodeIfPresent(String.self, forKey: .target)
        source = (try? container.decode(BackupSource.self, forKey: .source)) ?? .other
    }
}

/// Datastore PBS : c'est lui qui sature en premier quand la rétention dérive.
struct Datastore: Codable, Identifiable, Hashable, Sendable {
    var name: String
    var total: Double
    var used: Double
    var available: Double
    var percent: Double
    var estimatedFull: Date?
    var sourceHost: String
    var source: BackupSource

    var id: String { "\(sourceHost):\(name)" }

    enum CodingKeys: String, CodingKey {
        case name, total, used, available, percent, source
        case estimatedFull = "estimated_full"
        case sourceHost = "source_host"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        name = (try? container.decode(String.self, forKey: .name)) ?? "—"
        total = (try? container.decode(Double.self, forKey: .total)) ?? 0
        used = (try? container.decode(Double.self, forKey: .used)) ?? 0
        available = (try? container.decode(Double.self, forKey: .available)) ?? 0
        percent = (try? container.decode(Double.self, forKey: .percent)) ?? 0
        estimatedFull = container.decodeLooseDate(forKey: .estimatedFull)
        sourceHost = (try? container.decode(String.self, forKey: .sourceHost)) ?? "—"
        source = (try? container.decode(BackupSource.self, forKey: .source)) ?? .pbs
    }
}

/// Exécution de sauvegarde remontée par une source.
struct BackupTask: Codable, Identifiable, Hashable, Sendable {
    var upid: String?
    var type: String?
    var target: String?
    var user: String?
    var status: String?
    var started: Date?
    var ended: Date?
    var schedule: String?
    var next: Date?
    var sourceHost: String
    var source: BackupSource

    var id: String {
        upid ?? "\(sourceHost):\(target ?? type ?? "?"):\(started?.timeIntervalSince1970 ?? 0)"
    }

    enum CodingKeys: String, CodingKey {
        case upid, type, target, user, status, started, ended, schedule, next, source
        case sourceHost = "source_host"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        upid = try container.decodeIfPresent(String.self, forKey: .upid)
        type = try container.decodeIfPresent(String.self, forKey: .type)
        target = try container.decodeIfPresent(String.self, forKey: .target)
        user = try container.decodeIfPresent(String.self, forKey: .user)
        status = try container.decodeIfPresent(String.self, forKey: .status)
        started = container.decodeLooseDate(forKey: .started)
        ended = container.decodeLooseDate(forKey: .ended)
        schedule = try container.decodeIfPresent(String.self, forKey: .schedule)
        next = container.decodeLooseDate(forKey: .next)
        sourceHost = (try? container.decode(String.self, forKey: .sourceHost)) ?? "—"
        source = (try? container.decode(BackupSource.self, forKey: .source)) ?? .other
    }

    var label: String { target?.isEmpty == false ? target! : (type ?? "tâche") }

    /// Une tâche encore en cours n'est pas un échec : elle n'a simplement pas
    /// fini de rendre son verdict.
    var succeeded: Bool {
        guard let status, !status.isEmpty else { return true }
        return status == "OK" || status == "running"
    }

    var duration: Double? {
        guard let started, let ended, ended > started else { return nil }
        return ended.timeIntervalSince(started)
    }
}

/// Source injoignable au moment de l'agrégation.
struct SourceError: Codable, Identifiable, Hashable, Sendable {
    var host: String
    var kind: String
    var error: String

    var id: String { "\(host):\(kind)" }
}

/// Information non bloquante remontée par une source (Hyper Backup absent…).
struct SourceNotice: Codable, Identifiable, Hashable, Sendable {
    var host: String
    var kind: String
    var level: String
    var message: String

    var id: String { "\(host):\(message)" }

    var isWarning: Bool { level == "warning" }
}

/// Réponse de `GET /protection/sources` : sources déclarées et seuils du moteur.
struct ProtectionSources: Codable, Sendable {
    var sources: [Source]
    var hasPBS: Bool
    var thresholds: Thresholds

    enum CodingKeys: String, CodingKey {
        case sources, thresholds
        case hasPBS = "has_pbs"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        sources = (try? container.decode([Source].self, forKey: .sources)) ?? []
        hasPBS = (try? container.decode(Bool.self, forKey: .hasPBS)) ?? false
        thresholds = (try? container.decode(Thresholds.self, forKey: .thresholds)) ?? Thresholds()
    }

    struct Source: Codable, Identifiable, Hashable, Sendable {
        let id: Int
        var name: String
        var kind: String
        var address: String?
        var status: String?

        var hostKind: HostKind { HostKind(rawValue: kind) ?? .generic }
    }

    struct Thresholds: Codable, Sendable {
        var staleHours: Double = 36
        var criticalHours: Double = 168
        var storeWarn: Double = 80
        var storeCritical: Double = 92

        enum CodingKeys: String, CodingKey {
            case staleHours = "stale_hours"
            case criticalHours = "critical_hours"
            case storeWarn = "store_warn"
            case storeCritical = "store_crit"
        }

        init() {}
    }

    /// Le PBS qui détient un datastore donné, pour ouvrir ses instantanés.
    func pbsHost(named name: String) -> Source? {
        sources.first { $0.kind == "pbs" && $0.name == name }
    }
}

/// Un instantané d'un datastore PBS — `GET /protection/{host}/snapshots/{store}`.
struct PBSSnapshot: Codable, Identifiable, Hashable, Sendable {
    var store: String
    var backupType: String?
    var backupID: String?
    var time: Date?
    var size: Double
    var owner: String?
    var isProtected: Bool
    var verified: String?
    var comment: String?
    var files: Int

    var id: String { "\(store):\(backupType ?? "")/\(backupID ?? "")@\(time?.timeIntervalSince1970 ?? 0)" }

    enum CodingKeys: String, CodingKey {
        case store, time, size, owner, verified, comment, files
        case backupType = "backup_type"
        case backupID = "backup_id"
        case isProtected = "protected"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        store = (try? container.decode(String.self, forKey: .store)) ?? ""
        backupType = try container.decodeIfPresent(String.self, forKey: .backupType)
        // PBS numérote les invités : `backup-id` sort en nombre pour une VM.
        backupID = (try? container.decode(String.self, forKey: .backupID))
            ?? (try? container.decode(Int.self, forKey: .backupID)).map { String($0) }
        time = container.decodeLooseDate(forKey: .time)
        size = (try? container.decode(Double.self, forKey: .size)) ?? 0
        owner = try container.decodeIfPresent(String.self, forKey: .owner)
        isProtected = (try? container.decode(Bool.self, forKey: .isProtected)) ?? false
        verified = try container.decodeIfPresent(String.self, forKey: .verified)
        comment = try container.decodeIfPresent(String.self, forKey: .comment)
        files = (try? container.decode(Int.self, forKey: .files)) ?? 0
    }

    var group: String { "\(backupType ?? "?")/\(backupID ?? "?")" }
    var isVerified: Bool { verified == "ok" }
}

// MARK: - Dates hétérogènes

extension KeyedDecodingContainer {
    /// Les sources de sauvegarde datent chacune à leur façon : PBS et vzdump en
    /// époque Unix, DSM tantôt en époque tantôt en chaîne. Le décodeur global ne
    /// sait lire que des chaînes, d'où cette lecture tolérante.
    func decodeLooseDate(forKey key: Key) -> Date? {
        if let seconds = try? decodeIfPresent(Double.self, forKey: key) {
            // 0 vaut « jamais » côté PBS comme côté DSM.
            return seconds > 0 ? Date(timeIntervalSince1970: seconds) : nil
        }
        if let text = try? decodeIfPresent(String.self, forKey: key), !text.isEmpty {
            return DateParsing.parse(text)
        }
        return nil
    }
}
